docs/src/design/changing_the_primal.md

```diff
@@ -0,0 +1,516 @@
+```
+```julia
+y = sin(x)
+pullback_at(::typeof(sin), x, y, ȳ) = ȳ * cos(x)
+```
+`pullback_at` uses the primal input `x`, and the sensitivity being pulled back `ȳ`.
+
+```@raw html
+</details>
+```
+```@raw html
+<details><summary>Example for the logistic sigmoid</summary>
+```
+
+```julia
+σ(x) = 1/(1 + exp(-x))  # = exp(x)/(1+exp(x))
+y = σ(x)
+pullback_at(::typeof(σ), x, y, ȳ) = ȳ * y * σ(-x)  # i.e. ȳ * σ(x) * σ(-x)
+```
+Notice that here we are in the `pullback_at` not only using `x` but also `y` the primal output.
+This is a nice bit of symmetry that shows up around `exp`.
+```@raw html
+</details>
+```
+
+Now let's consider why we implement `rrule`s in the first place.
+One key reason [^3] is to allow us to insert our domain knowledge to do better than the AD would do just by breaking everything down into `+`, `*`, etc.
+What insights do we have about `sin` and `cos`?
+What about using `sincos`?
+```@raw html
+<details><summary>Example for `sin`</summary>
+```
+```julia
+julia> using BenchmarkTools
+
+julia> @btime sin(x) setup=(x=rand());
+  3.838 ns (0 allocations: 0 bytes)
+
+julia> @btime cos(x) setup=(x=rand());
+  4.795 ns (0 allocations: 0 bytes)
+
+julia> 3.838 + 4.795
+8.633
+```
+vs computing both together:
+```julia
+julia> @btime sincos(x) setup=(x=rand());
+  6.028 ns (0 allocations: 0 bytes)
+```
+```@raw html
+</details>
+```
+
+What about the logistic sigmoid?
+We note that the two values we need are `σ(x)` and `σ(-x)`
+If we write these as:
+$\sigma(x) = \frac{e^x}{1+e^x}$ and
+$\sigma(-x) = \frac{1}{1+e^x}$
+then we see they have the common term $e^x$.
+`exp(x)` is a much more expensive operation than `+` and `/`.
+So we can save time, if we can reuse that `exp(x)`.
+```@raw html
+<details><summary>Example for the logistic sigmoid</summary>
+```
+If we have to computing separately:
+```julia
+julia> @btime 1/(1+exp(x)) setup=(x=rand());
+  5.622 ns (0 allocations: 0 bytes)
+
+julia> @btime 1/(1+exp(-x)) setup=(x=rand());
+  6.036 ns (0 allocations: 0 bytes)
+
+julia> 5.622 + 6.036
+11.658
+```
+
+vs reusing `exp(x)`:
+```julia
+julia> @btime exp(x) setup=(x=rand());
+  5.367 ns (0 allocations: 0 bytes)
+
+julia> @btime ex/(1+ex) setup=(ex=exp(rand()));
+  1.255 ns (0 allocations: 0 bytes)
+
+julia> @btime 1/(1+ex) setup=(ex=exp(rand()));
+  1.256 ns (0 allocations: 0 bytes)
+
+julia> 5.367 + 1.255 + 1.256
+7.878
+```
+```@raw html
+</details>
+```
+
+So we are talking about a 30-40%[^4] speed-up from these optimizations.
+It's faster to  compute `sin` and `cos` at the same time via `sincos` than it is to compute them one after the other.
+And it is faster to reuse the `exp(x)` in computing `σ(x)` and `σ(-x)`.
+How can we incorporate this insight into our system?
+We know we can compute both of these in the primal, because they only depend on `x` --- we don't need to know `ȳ`.
+But there is nowhere to put it that is accessible both to the primal pass and the gradient pass code.
```

Comment:
```suggestion
We know we can compute both of these in the primal, because they only depend on `x` --- we don't need to know `ȳ`.
But there is nowhere to put it that is accessible both to the primal pass and the gradient pass code.
```
```suggestion
We know we can compute both of these in the primal — because they only depend on `x` and not on `ȳ` — but there is nowhere to put them that is accessible both to the primal pass and the gradient pass code.
```